Missing @Nullable in AsyncMap, Context, Vertx and WorkerExecutor

tsegismont opened the pull request:
Fixes #2434

Improves Vert.x RxJava2 API (returns Maybe instead of Single)

With executeBlocking, the future provided to the blocking handler could be completed with null.

In AsyncMap, there are a few cases (get, remove, replace, putIfAbsent) when the result can be null as well.

vietj commented on src/main/java/io/vertx/core/Context.java:
```diff
@@ -116,15 +116,15 @@ static boolean isOnVertxThread() {
    * @param <T> the type of the result
    */
-  <T> void executeBlocking(Handler<Future<T>> blockingCodeHandler, Handler<AsyncResult<T>> resultHandler);
+  <T> void executeBlocking(Handler<Future<T>> blockingCodeHandler, Handler<AsyncResult<@Nullable T>> resultHandler);
```
it needs also to be `Handler<Future<@Nullable T>>`

vietj commented on src/main/java/io/vertx/core/Vertx.java:
```diff
@@ -519,12 +519,12 @@ static void clusteredVertx(VertxOptions options, Handler<AsyncResult<Vertx>> res
    * @param <T> the type of the result
    */
-  <T> void executeBlocking(Handler<Future<T>> blockingCodeHandler, boolean ordered, Handler<AsyncResult<T>> resultHandler);
+  <T> void executeBlocking(Handler<Future<T>> blockingCodeHandler, boolean ordered, Handler<AsyncResult<@Nullable T>> resultHandler);
```
it needs also to be `Handler<Future<@Nullable T>>`

vietj commented on src/main/java/io/vertx/core/Vertx.java:
```diff
@@ -519,12 +519,12 @@ static void clusteredVertx(VertxOptions options, Handler<AsyncResult<Vertx>> res
    * Like {@link #executeBlocking(Handler, boolean, Handler)} called with ordered = true.
    */
-  <T> void executeBlocking(Handler<Future<T>> blockingCodeHandler, Handler<AsyncResult<T>> resultHandler);
+  <T> void executeBlocking(Handler<Future<T>> blockingCodeHandler, Handler<AsyncResult<@Nullable T>> resultHandler);
```
it needs also to be `Handler<Future<@Nullable T>>`

vietj commented on src/main/java/io/vertx/core/WorkerExecutor.java:
```diff
@@ -47,12 +48,12 @@
    * @param <T> the type of the result
    */
-  <T> void executeBlocking(Handler<Future<T>> blockingCodeHandler, boolean ordered, Handler<AsyncResult<T>> resultHandler);
+  <T> void executeBlocking(Handler<Future<T>> blockingCodeHandler, boolean ordered, Handler<AsyncResult<@Nullable T>> resultHandler);
```
it needs also to be `Handler<Future<@Nullable T>>`

vietj commented on src/main/java/io/vertx/core/WorkerExecutor.java:
```diff
@@ -47,12 +48,12 @@
    * Like {@link #executeBlocking(Handler, boolean, Handler)} called with ordered = true.
    */
-  default <T> void executeBlocking(Handler<Future<T>> blockingCodeHandler, Handler<AsyncResult<T>> resultHandler) {
+  default <T> void executeBlocking(Handler<Future<T>> blockingCodeHandler, Handler<AsyncResult<@Nullable T>> resultHandler) {
```
it needs also to be `Handler<Future<@Nullable T>>`

vietj commented on src/main/java/io/vertx/core/shareddata/AsyncMap.java:
```diff
@@ -81,15 +82,15 @@
    * @param completionHandler  the handler
    */
-  void putIfAbsent(K k, V v, long ttl, Handler<AsyncResult<V>> completionHandler);
+  void putIfAbsent(K k, V v, long ttl, Handler<AsyncResult<@Nullable V>> completionHandler);
```
the second param `V` should also be `@Nullable`

tsegismont replied:
No, null values are forbidden in `AsyncMap`.

vietj replied:
ah ok makese sense then, perhaps we should say it in the javadoc as well then

vietj commented on src/main/java/io/vertx/core/shareddata/AsyncMap.java:
```diff
@@ -108,7 +109,7 @@
    * @param resultHandler  the result handler will be passed the previous value
    */
-  void replace(K k, V v, Handler<AsyncResult<V>> resultHandler);
+  void replace(K k, V v, Handler<AsyncResult<@Nullable V>> resultHandler);
```
the second param `V` should also be `@Nullable`

tsegismont replied:
No, null values are forbidden in AsyncMap.

vietj replied:
ah ok makese sense then, perhaps we should say it in the javadoc as well then

tsegismont commented:
@vietj good point, I only thought about RxJava2, but alt langs implementation like Scala could also benefit.

Nevertheless, the build fails with  a `Handler<Future<@Nullable T>>` parameter. I will create an issue in codegen. Can I go ahead and merge this one in the meantime?

vietj commented:
+1 for issue in codegen, however I think we shall keep this issue open until we fix it in codegen

tsegismont commented:
@vietj all the changes are in

tsegismont commented:
@vietj I just rebased and fixed this one after closing vert-x3/vertx-codegen#180

Can you please take another look (just a couple of classes and a few lines changes)? Thanks

vietj commented:
ideally we should declare

```
<@Nullable T> void executeBlocking(Future<T> fut, Handler<AsyncResult<T>> handler)
```

but codegen does not support this declaration (yet)

so for now it's fine the accept the PR as is.
